New package: Reactant v0.1.0

JuliaRegistrator opened the pull request:
- Registering package: Reactant
- Repository: https://github.com/EnzymeAD/Reactant.jl
- Created by: @wsmoses
- Version: v0.1.0
- Commit: f06a9436b7366b1f30b3b9862f933688bf6bcbfd
- Reviewed by: @wsmoses
- Reference: https://github.com/EnzymeAD/Reactant.jl/commit/f06a9436b7366b1f30b3b9862f933688bf6bcbfd#commitcomment-142087028
<!-- bf0c69308befbd3ccf2cc956ac8a46712550b79fc9bfb5e4edf8f833f05f4c18b06eddad8845b45beb9f45c2b8020dd6568be1411e8c6f15a8cf2d0fba5d1261ef437471b5ad9870534f2d58a44988aa0a5e4184222e9e8fc6f81689f7acde7385fe916052258ffafe76398a40ca6d862eab4595e82ccba37b4b4ba1160822aea94e9f64a64c8a48743b216249a5677b48f05ef5d7c9ab7b62881d11bf1e219fab9fa6a41406bc71360ea436d19833dd4b797c6d57c6182c4d0c438fc26e6a2bfe910bdad952ad12ac5af28454e36121a54788eb254cff1e4ae99b0ec2979a28 -->

goerz commented:
The package is also missing documentation

wsmoses commented:
Yeah doc construction is failing until we get a bump on yggdrassil for reactant_jll to 0.0.6 since doc CI doesn't have a GPU

goerz commented:
Some description, context, and usage example in the README would already go a long way! 😉

wsmoses commented:
Presently the readme is intentionally bare as end users should not use the package. I'm registering it presently to make it easier for developers to build it and perform smoke tests in meta packages to see what needs to be built (e.g. it provides functionality for other packages in the ecosystem).

You're welcome to browse our test suite for example usage, including https://github.com/EnzymeAD/Reactant.jl/blob/main/test/basic.jl and https://github.com/EnzymeAD/Reactant.jl/blob/main/test/nn.jl.

wsmoses commented:
@goerz as I'm realizing that one needs to include [noblock] for this to get unstuck once I finish getting our docs to build, can you add that to your comments?

internal developer documentation will be online as soon as yggdrassil merges a new jll (https://github.com/JuliaPackaging/Yggdrasil/pull/8676) and we adapt docs-CI to not require a GPU.

goerz commented:
> Presently the readme is intentionally bare as end users should not use the package.

Huh. Is that a temporary situation? Because for a non-user-facing package, I'm not sure if I'm comfortable with "Reactant". That's the kind of nice whimsical package name that would make a nice pun in the context of Enzyme. But if it's just for internal use, such memorable names would be better kept open for other projects that get some mileage out of them, IMO.

It also doesn't seem to me that empty READMEs are ever okay, even from seasoned developers or well-established orgs like Enzyme. Maybe *especially* not: the newcomers we tell to add some minimal information to their README would rightfully argue "these high-profile orgs get away with registering undocumented packages". So I'm very much in favor of keeping *everyone* to the same ([very minimal](https://discourse.julialang.org/t/when-does-it-make-sense-to-post-a-new-package-to-the-julia-repository/108701/4)) standards. Which doesn't include a fully built documentation, but it does include a minimal README.

Even for an internal package, there's no harm in saying in the README "This package provides the following internal functionality in the context of this other user-facing package. Please see the documentation of the main project."

> Can you add [noblock] to your comments?

I would, but I'm actually not sure that this registration should go through in its current state. If there was enough context to understand the justification for the name, and with a minimal README to explain what this does and that more complete documentation is forthcoming, sure.

Well, you still have three days to fill in a bit more context and for me to unblock, or otherwise, maybe this needs a bit more community discussion on Slack.

wsmoses commented:
[noblock]

Yeah it is destined for end user use. In particular, copying a brief description I made in https://github.com/EnzymeAD/Enzyme.jl/issues/805#issuecomment-2106061410 

"""
Scheduling. Like I said at the top this is key to good performance, and incidentally distinct from AD. I have started playing with a repo Reactant.jl (https://github.com/EnzymeAD/Reactant.jl) which aims to resolve this. It can take a julia function and compile it into MLIR and run fancy optimizations on top of it, including using EnzymeMLIR for AD, and create relevant executables for CPU/GPU/TPU via XLA. It is very much in progress, but nevertheless a problem outside Enzyme now.
"""

And oh, my apologies I didn't realize that registration required docs. This is my first time registering a non-jll package, and that process seemed pretty cookie cutter.

In any cases I added a (reasonably cautious) README here, if that looks good to you @goerz : https://github.com/EnzymeAD/Reactant.jl

wsmoses commented:
and yes the name choice is explicit. Whereas Enzyme deals with differentiaiton. Reactant is an "execution engine". They are both separately useful for reasons specified in each doc in their own right, but they do work well together (and a reactant is an input to an Enzyme)

goerz commented:
[noblock] Looks great! All good!

wsmoses commented:
and yeah part of why I decided to register presently is that I learned people were starting to test it out in non-general registries, which seemed like a good reason to register it (in spite of my heavy warnings about potential semantic assurances).

[noblock]

goerz commented:
[noblock] “Stable” is very much not a prerequisite for registration. Only “has some functionality” and “a general Julia user can have an idea of what the package is for and how to use it” (aka documentation / README). So this seems absolutely fine to put in General at this point.

goerz commented:
[noblock] (of course you still have to fix the compat issues and whatever is causing the package not to load)

JuliaRegistrator committed:
New package: Reactant v0.1.0

UUID: 3c362404-f566-11ee-1572-e11a4b42c853
Repo: https://github.com/EnzymeAD/Reactant.jl.git
Tree: a24822bc0f4cdf6a1384f7e869885473bc410e09

Registrator tree SHA: 17aec322677d9b81cdd6b9b9236b09a3f1374c6a
